feat: log most recent API calls as `recent-bigframes-api-xx` labels on BigQuery jobs

ashleyxuu opened the pull request:
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery-dataframes/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> 🦕

ashleyxuu committed:
docs: link to ML.EVALUATE BQML page for score() methods

ashleyxuu committed:
feat: label query job with bigframes-api-xx using decorator

ashleyxuu committed:
reorganize the commit

ashleyxuu committed:
Merge branch 'main' into ashleyxu-add-api-methods

ashleyxuu commented:
FYI, the logic for creating labels doesn't change as https://github.com/googleapis/python-bigquery-dataframes/pull/55. I just added the `utils/log_adatper` so we can add to the `api_methods` list globally when calling the methods. Therefore, we don't need to pass down `api_name` params in each layer.

shobsi committed:
test: Log slowest tests durations (#146)

Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://togithub.com/googleapis/python-bigquery-dataframes/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> 🦕

ashleyxuu committed:
docs: link to ML.EVALUATE BQML page for score() methods (#137)

Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://togithub.com/googleapis/python-bigquery-dataframes/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> 🦕

tswast commented on bigframes/utils/log_adapter.py:
```diff
@@ -0,0 +1,32 @@
+    def wrapper(*args, **kwargs):
+        api_method_name = str(method.__name__)
+        if not api_method_name.startswith("__"):
+            add_api_method(api_method_name)
```
Arun't you missing return value here?

ashleyxuu replied:
Done.

ashleyxuu committed:
feat: populate ibis version in user agent (#140)

Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://togithub.com/googleapis/python-bigquery-dataframes/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> 🦕

tswast committed:
fix: don't override the global logging config (#138)

Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://togithub.com/googleapis/python-bigquery-dataframes/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> 🦕

milkshakeiii committed:
fix: use indexee's session for loc listlike cases (#152)

Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://togithub.com/googleapis/python-bigquery-dataframes/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> 🦕

TrevorBergeron committed:
feat: add pandas.qcut (#104)

Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://togithub.com/googleapis/python-bigquery-dataframes/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> 🦕

TrevorBergeron committed:
feat: add unstack to series, add level param (#115)

Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://togithub.com/googleapis/python-bigquery-dataframes/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> 🦕

tswast committed:
feat: add `DataFrame.to_pandas_batches()` to download large `DataFrame` objects (#136)

Builds on https://togithub.com/googleapis/python-bigquery-dataframes/pull/132

Towards internal issue 280662868
🦕

orrbradford committed:
fix: resolve plotly rendering issue by using ipython html for job pro… (#134)

…gress messages

Fixes bug that was preventing plotly rendering to show after the progress bar. Original ipywidgets implementation isn't necessary for basic opening of urls

Screen recording: https://togithub.com/googleapis/python-bigquery-dataframes/assets/15842009/5225ce05-117a-4808-9ff0-cb2c3aaf3a40

Internal bug: b/297062404

TrevorBergeron committed:
refactor: ArrayValue is now a tree that defers conversion to ibis (#110)

Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://togithub.com/googleapis/python-bigquery-dataframes/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> 🦕

milkshakeiii committed:
fix: fix bug with column names under repeated column assignment (#150)

Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://togithub.com/googleapis/python-bigquery-dataframes/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> 🦕

shobsi committed:
test: refactor remote function tests (#147)

This changes moves the tests that deploy cloud function to large remote function tests, and the tests that do not make call to bigquery service to unit tests.

Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://togithub.com/googleapis/python-bigquery-dataframes/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> 🦕

TrevorBergeron committed:
feat: add dataframe melt (#116)

Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://togithub.com/googleapis/python-bigquery-dataframes/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> 🦕

ashleyxuu committed:
docs: add artithmetic df sample code (#153)

* docs: add artithmetic df sample code

* fix: address comments

shobsi committed:
feat: Implement operator `@` for `DataFrame.dot` (#139)

Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://togithub.com/googleapis/python-bigquery-dataframes/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes b/297502513 🦕

ashleyxuu committed:
fix: fix typo and address comments

ashleyxuu committed:
Merge branch 'main' into ashleyxu-add-api-methods

tswast committed:
Merge branch 'main' into ashleyxu-add-api-methods

tswast commented on tests/unit/session/test_io_bigquery.py:
```diff
@@ -19,15 +19,122 @@
+    expected_dict = {
+        "bigframes-api": "read_pandas",
+        "source": "bigquery-dataframes-temp",
+        "bigframes-api-2": "head",
```
I don't really understand why the count starts at 2. I would think an existing "bigframes-api" label would be totally opaque to the logic we use to add the labels.

tswast commented on bigframes/utils/log_adapter.py:
```diff
@@ -0,0 +1,37 @@
+    @functools.wraps(method)
+    def wrapper(*args, **kwargs):
+        api_method_name = str(method.__name__)
+        if not api_method_name.startswith("_"):
```
This isn't entirely accurate. We do want to add when it starts and ends with two `__` characters. Those are the "dunder" methods and how Python overrides operators.

tswast commented on bigframes/session/_io/bigquery.py:
```diff
@@ -17,15 +17,45 @@
+    job_configs_labels: Optional[Dict[str, str]],
+    api_methods: Sequence[str],
+) -> Dict[str, str]:
+    # If there is no label set
+    if job_configs_labels is None:
+        labels = {}
+        label_values = list(api_methods)
+    else:
+        labels = job_configs_labels.copy()
+        cur_labels_len = len(job_configs_labels)
+        api_methods_len = len(api_methods)
+        # If the total number of labels is under the limit of labels count
+        if cur_labels_len + api_methods_len <= MAX_LABELS_COUNT:
+            label_values = list(api_methods)
+        # We capture the latest label if it is out of the length limit of labels count
+        else:
+            added_api_len = cur_labels_len + api_methods_len - MAX_LABELS_COUNT
+            label_values = list(api_methods)[-added_api_len:]
+
+    for i, label_value in enumerate(label_values):
+        if job_configs_labels is not None:
+            label_key = "bigframes-api-" + str(i + len(job_configs_labels))
+        else:
+            label_key = "bigframes-api-" + str(i)
+        labels[label_key] = label_value
+    return labels
```
This all seems way more complicated than it needs to be. Some functional-style code could make this a lot more understandable. Note: You'll need to import `itertools` for `chain`, which is a built-in module.

```suggestion
    # If there is no label set
    if job_configs_labels is None:
        job_configs_labels = {}

    labels = list(itertools.chain(
        job_config_labels.keys(),
        (f"bigframes-api-{i}" for i in range(len(api_methods)))
    ))
    values = list(itertools.chain(job_config_labels.values(), api_methods))
    return dict(zip(labels[:MAX_LABELS_COUNT], values[:MAX_LABELS_COUNT]))
```

tswast commented on tests/unit/session/test_io_bigquery.py:
```diff
@@ -19,15 +19,122 @@
+    assert labels == expected_dict
+
+
+def test_create_job_configs_labels_length_limit_met():
```
You're missing a test where the length limit is met and there aren't any current labels.

ashleyxuu committed:
fix: address comments

tswast commented on bigframes/core/groupby/__init__.py:
```diff
@@ -30,9 +30,11 @@
 import bigframes.operations as ops
 import bigframes.operations.aggregations as agg_ops
 import bigframes.series as series
+from bigframes.utils import log_adapter
```
Please remove the "utils" package. "core" already serves the same purpose and communicates that the modules are internal.

tswast replied:
go/no-utils

ashleyxuu committed:
Remove utils folder and refactor it in core directory

ashleyxuu committed:
Remove utils folder and refactor it in core directory

ashleyxuu committed:
Merge branch 'main' into ashleyxu-add-api-methods

ashleyxuu committed:
Merge remote-tracking branch 'origin/ashleyxu-add-api-methods' into ashleyxu-add-api-methods

ashleyxuu committed:
Merge branch 'main' into ashleyxu-add-api-methods

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

ashleyxuu committed:
fix merge conflicts

ashleyxuu committed:
Merge remote-tracking branch 'origin/ashleyxu-add-api-methods' into ashleyxu-add-api-methods

tswast commented on bigframes/core/__init__.py:
```diff
@@ -39,6 +40,8 @@
 ORDER_ID_COLUMN = "bigframes_ordering_id"
 PREDICATE_COLUMN = "bigframes_predicate"
 
+__all__ = ["class_logger", "method_logger"]
```
Let's remove this. It doesn't really make sense to re-export these methods just for a slightly smaller import. If we were to have it, we'd want to include other classes as well, such as ArrayValue.

```suggestion
```

tswast commented on bigframes/core/log_adapter.py:
```diff
@@ -0,0 +1,38 @@
+import functools
```
You're missing a license header.

tswast commented on bigframes/core/log_adapter.py:
```diff
@@ -0,0 +1,38 @@
+from loguru import logger
+
+_log_file_path = None
+_logger = logger
```
This variable appears to be unused.

tswast commented on bigframes/core/log_adapter.py:
```diff
@@ -0,0 +1,38 @@
+
+from loguru import logger
+
+_log_file_path = None
```
This variable appears to be unused.

tswast commented on bigframes/core/log_adapter.py:
```diff
@@ -0,0 +1,38 @@
+
+_log_file_path = None
+_logger = logger
+_api_methods = []
```
I'm confused by this. Wouldn't we want to store the `_api_methods` on the individual objects. What happens when we have two different objects? I don't think we want to log all methods ever in the API request, just the ones used to generate the object.

tswast replied:
Need to protect this with a lock.

tswast commented on bigframes/core/log_adapter.py:
```diff
@@ -0,0 +1,38 @@
+    for attr_name, attr_value in decorated_cls.__dict__.items():
+        if callable(attr_value):
+            setattr(decorated_cls, attr_name, method_logger(attr_value))
+    return decorated_cls
```
This would be where we'd override the initializer, I think to create an `_api_methods` attribute, potentially combining the `_api_methods` from objects passed into the constructor.

tswast commented on bigframes/core/log_adapter.py:
```diff
@@ -0,0 +1,38 @@
+import functools
+
+from loguru import logger
```
Doesn't look like we actually use loguru. Let's remove that dependency if so.

tswast commented on third_party/bigframes_vendored/cpython/Lib/functools.py:
```diff
@@ -0,0 +1,101 @@
+"""functools.py - Tools for working with functions and callable objects
```
Why are we vendoring this? Just use the one built-in to Python, right?

tswast commented on bigframes/session/_io/bigquery.py:
```diff
@@ -30,9 +31,28 @@
+    labels = list(
+        itertools.chain(
+            job_configs_labels.keys(),
+            (f"bigframes-api-{i}" for i in range(len(api_methods))),
```
I wonder if we actually want to reverse the `api_methods` list, or make sure we always add the latest API to the front?

tswast replied:
Since these might not actually relate to the actual job being run, let's rename the labels to `recent-bigframes-api-{i}`.

tswast commented on bigframes/core/log_adapter.py:
```diff
@@ -0,0 +1,38 @@
+
+def add_api_method(method: str):
+    global _api_methods
+    _api_methods.append(method)
```
Make sure we use a lock before modifying this. Push to the front and keep the list of APIs less than our maximum length.

tswast commented on bigframes/session/__init__.py:
```diff
@@ -1506,8 +1508,14 @@ def _start_query(
         """
         Starts query job and waits for results.
         """
+        api_methods = log_adapter._api_methods
         job_config = self._prepare_job_config(job_config)
+        job_config.labels = bigframes_io.create_job_configs_labels(
+            job_configs_labels=job_config.labels, api_methods=api_methods
+        )
         query_job = self.bqclient.query(sql, job_config=job_config)
+        # Clear out the global api logger
+        log_adapter._api_methods = []
```
This all needs to be protected by a lock. Alternatively, let's make a `log_adapter.get_and_reset_api_methods()` that locks correctly and returns a copy of the previous list of api methods.

ashleyxuu committed:
commit the conflicts

ashleyxuu committed:
redesign the log adapter

ashleyxuu committed:
resolve conflicts and merge remote-tracking branch 'origin/main' into ashleyxu-add-api-methods

ashleyxuu committed:
Make the global _api_methods and lock threads

ashleyxuu committed:
Make the global _api_methods and lock threads

ashleyxuu committed:
Merge branch 'main' into ashleyxu-add-api-methods

ashleyxuu committed:
merge conflicts

tswast commented on bigframes/core/log_adapter.py:
```diff
@@ -0,0 +1,65 @@
+        # Track regular and "dunder" methods
+        if api_method_name.startswith("__") or not api_method_name.startswith("_"):
+            add_api_method(api_method_name)
+        try:
+            result = method(*args, **kwargs)
+            return result
+        except Exception as e:
+            raise e
```
Since we aren't doing anything with the exception, we don't really need to re-raise, right? We can add this back in future if we want to start logging client-side exceptions.

```suggestion
        return method(*args, **kwargs)
```

tswast commented on bigframes/session/_io/bigquery.py:
```diff
@@ -17,17 +17,37 @@
+    job_configs_labels: Optional[Dict[str, str]],
+    api_methods: Sequence[str],
+) -> Dict[str, str]:
+    # If there is no label set
```
This comment is just describing _what_ the code does, not _why_. See: go/comments

```suggestion
```

tswast commented on tests/unit/core/test_log_adapter.py:
```diff
@@ -0,0 +1,44 @@
+import pytest
```
Need license header.

ashleyxuu committed:
address comments

ashleyxuu committed:
address comments

ashleyxuu committed:
Merge remote-tracking branch 'origin/ashleyxu-add-api-methods' into ashleyxu-add-api-methods

ashleyxuu committed:
fix error

ashleyxuu committed:
fix None job_config error

tswast commented on tests/unit/test_compute_options.py:
```diff
@@ -26,5 +26,6 @@ def test_maximum_bytes_option():
             _, _, kwargs = call
             num_query_calls += 1
-            assert kwargs["job_config"].maximum_bytes_billed == 10000
+            if "job_config" in kwargs:
+                assert kwargs["job_config"].maximum_bytes_billed == 10000
```
I find this concerning. Please revert.

ashleyxuu committed:
address comments

ashleyxuu committed:
Merge branch 'main' into ashleyxu-add-api-methods
